backends/utils/history_server.go

```diff
@@ -0,0 +1,526 @@
+	defaultWriteMonitoringLogsTimeout   = 10 * time.Second
+	defaultPendingLogsBatchSize         = 100
+	defaultLogsFolderPath               = "/monitoring/frames"
+	defaultLogsContainer                = "users"
+	defaultMaxBytesInNginxRequest       = 5 * 1024 * 1024 // 5MB
+	maxLogsInMessage                    = 2
+
+	queryType = "query"
+)
+
+type HistoryEntry struct {
+	ActionType     string
+	UserName       string
+	BackendName    string
+	TableName      string
+	Container      string
+	ActionDuration time.Duration
+	StartTime      time.Time
+
+	AdditionalData map[string]string
+}
+
+type HistoryServer struct {
+	logger    logger.Logger
+	logs      []HistoryEntry
+	requests  chan HistoryEntry
+	container v3io.Container
+	config    *frames.Config
+	isActive  bool
+
+	WriteMonitoringLogsTimeout     time.Duration
+	PendingLogsBatchSize           int
+	LogsFolderPath                 string
+	LogsContainer                  string
+	MaxBytesInNginxRequest         int
+	HistoryFileDurationSecondSpans int64
+	HistoryFileNum                 int
+}
+
+func NewHistoryServer(logger logger.Logger, cfg *frames.Config) *HistoryServer {
+	mon := HistoryServer{
+		logger:   logger,
+		requests: make(chan HistoryEntry, 100),
+		isActive: false,
+		config:   cfg}
+
+	mon.initDefaults()
+
+	mon.createDefaultV3ioClient()
+	if mon.isActive {
+		mon.Start()
+		mon.StartEvictionTask()
+	}
+	return &mon
+}
+
+func (m *HistoryServer) initDefaults() {
+	m.WriteMonitoringLogsTimeout = defaultWriteMonitoringLogsTimeout
+	if m.config.WriteMonitoringLogsTimeoutSeconds != 0 {
+		m.WriteMonitoringLogsTimeout = time.Duration(m.config.WriteMonitoringLogsTimeoutSeconds) * time.Second
+	}
+
+	m.PendingLogsBatchSize = defaultPendingLogsBatchSize
+	if m.config.PendingLogsBatchSize != 0 {
+		m.PendingLogsBatchSize = m.config.PendingLogsBatchSize
+	}
+
+	m.LogsFolderPath = defaultLogsFolderPath
+	if m.config.LogsFolderPath != "" {
+		m.LogsFolderPath = m.config.LogsFolderPath
+	}
+
+	m.LogsContainer = defaultLogsContainer
+	if m.config.LogsContainer != "" {
+		m.LogsContainer = m.config.LogsContainer
+	}
+
+	m.MaxBytesInNginxRequest = defaultMaxBytesInNginxRequest
+	if m.config.MaxBytesInNginxRequest != 0 {
+		m.MaxBytesInNginxRequest = m.config.MaxBytesInNginxRequest
+	}
+
+	m.HistoryFileDurationSecondSpans = defaultHistoryFileDurationInSeconds
+	if m.config.HistoryFileDurationHourSpans != 0 {
+		m.HistoryFileDurationSecondSpans = m.config.HistoryFileDurationHourSpans * 3600
+	}
+
+	m.HistoryFileNum = defaultHistoryFileNum
+	if m.config.HistoryFileNum != 0 {
+		m.HistoryFileNum = m.config.HistoryFileNum
+	}
+}
+
+func (m *HistoryServer) getCurrentLogFileName() string {
+	// Round current time to the nearest log time-bucket
+	currentTimeInSeconds := time.Now().Unix()
+	currentTimeBucket := m.HistoryFileDurationSecondSpans * (currentTimeInSeconds / m.HistoryFileDurationSecondSpans)
+	return fmt.Sprintf("%v/%v.json", m.LogsFolderPath, currentTimeBucket)
```

Comment:
Might be more user friendly to use an ISO timestamp than unix time for the file name.